lib/shipit/line_buffer.rb

```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+module Shipit
+  class LineBuffer
+    SEPARATOR = "\n"
+
+    def initialize(value = "")
+      @value = value
+    end
+
+    def buffer(text, &block)
+      @value += text
+      whole_lines.each(&block).tap { flush }
+    end
+
+    private
+
+    def whole_lines
+      whole? ? lines : lines[0..-2]
+    end
+
+    def flush
+      whole? ? clear : @value = lines.last
+    end
+
+    def whole?
+      @value.end_with?(SEPARATOR)
```

Comment:
What will happen if the output does not contain a trailing newline (try `echo -n hello` in your shell). Will we lose the buffered output?

Reply:
It should probably be entirely flushed when the task ends.

Reply:
I thought the command output always ended with a new line. If it's interrupted, the half line would disappear with the line buffer when it's garbage collected. One option here if you think we're going to get important half lines is to flush it entirely when the task ends.

Reply:
> if you think we're going to get important half lines

When a command crash, the last line is often the most important one.

Reply:
What's the best way to catch the task ending for any reason i.e. failed, aborted, success, etc?

Reply:
It happens here: https://github.com/Shopify/shipit-engine/blob/73897c9268235ceb03ce8e335ebe5d931297275c/app/jobs/shipit/perform_task_job.rb#L17-L33

So the various `report_*` methods. Might want to call some flush in them.

Reply:
So it looks like all cases apart from `complete` already write log messages with new lines.

Reply:
There are a few options I’m considering here. I’d like to know what you think.

1. Write a final log message in all cases, ensuring a final whole line is sent to the buffer. Currently, `complete` is the only case missing a final log message. All the other cases are covered as far as I can tell.
2. Add an `ensure` clause to `PerformTaskJob#run` that reports it as `finished` i.e. the job is done, whether it's completed, timed out, erred, or anything else. Then connect that to a method that resolves the buffer.
3. Resolve the buffer directly from each case in `PerformTaskJob#run` (all the `report_*` methods @casperisfine mentioned).
4. Resolve the buffer on an `after_transition` callback in `Task`.

Reply:
We should try to keep the command output contiguous. Option 4 is harder to reason about in terms of ordering, 1-3 all seem fine. Would adding an ensure block [here](https://github.com/Shopify/shipit-engine/blob/73897c9268235ceb03ce8e335ebe5d931297275c/app/jobs/shipit/perform_task_job.rb#L76) also work?

```ruby
def capture!
...
ensure
  @task.write('\n') # move this down from the end of the main block
  # or @task.output_line_buffer.flush 
  # or something something
end
```

Reply:
Thanks @DazWorrall! We can't consistently resolve the buffer by simply writing `\n` since we no longer ignore blank lines. I’m going forward with option 1 for now, since it's the simplest. If we feel like that's not right, I’m more than happy to revisit option 2 or something like that another day.